nav2_costmap_2d/src/costmap_2d.cpp

```diff
@@ -57,6 +59,38 @@ Costmap2D::Costmap2D(
   resetMaps();
 }
 
+Costmap2D::Costmap2D(const nav_msgs::msg::OccupancyGrid & map)
+: default_value_(FREE_SPACE)
+{
+  access_ = new mutex_t();
+  std::unique_lock<mutex_t> lock(*access_);
+
+  // fill local variables
+  size_x_ = map.info.width;
+  size_y_ = map.info.height;
+  resolution_ = map.info.resolution;
+  origin_x_ = map.info.origin.position.x;
+  origin_y_ = map.info.origin.position.y;
+
+  // create the costmap
+  costmap_ = new unsigned char[size_x_ * size_y_];
+
+  // fill the costmap with a data
+  int8_t data;
+  for (unsigned int it = 0; it < size_x_ * size_y_; it++) {
```

Comment:
Do you mean `base` and `multiplier`? These parameters will be used in `SpeedFilter` for the maximum speed calculation for further `nav2_msgs::msg::MaxSpeed` message publishing. This is not an attribute of costmap, so I do not think we need to add them here.

Reply:
Yeah, I agree that it might not be necessary for this particular layer, but where would you use it in this context? I guess I don't see a really natural place for it.

Reply:
Yep, they are only used by Speed Filter, in Keepout Filter they are not used.

Reply:
I'm asking where you would use that information when you eventually need to for the speed layer? Its not important for this PR, I'm just trying to understand where you will eventually put it when you need to. You should look at those values for this layer too, its useful information **and someone could change them to be non-identity transforms** so you should be at least accessing them

Reply:
There will be a new class called `SpeedFilter` inherited from `CostmapFilter` as well as keepout filter does. There will be set a new subscriber to `CostmapFilterInfo` and according callback where I will read these parameters from an incoming message and set `base_`/`multiplier_` private values of `SpeedFilter` class. Later in `SpeedFilter::process()` routine when robot is entering a speed-restriction zone, these values will be used to reduce robot speed by sending a message to DWB controller.

Regarding the reading of these values in `CostmapFilter` class: by initial design this class has no subscribers set and callbacks handling, since I thought it is filter's responsibility (filter plugin can decide: use or not use any specific subscription). So, currently there is not a place here we can read the message until the the subscription to the `CostmapFilterInfo.msg` will be handled inside `CostmapFilter` (I can do this, but currently not sure that this is the right step because of the reason above).

Reply:
I'll add `SpeedFilter` code in this PR as soon as I will sort feature tests out.

Reply:
Please don't add the speed filter to this PR. This PR has enough in it, we should add that in the next PR.

So you're storing the non-tranformed values and then only transforming them in `process()`? For the speed layer, I think that's an OK thing to do because we're only looking at 1 cell (where the robot is), but for any other costmap filter, that would be not good. If you have to iterate over some local window, that means every iteration, you're owning a multiplication and addition for each grid cell. And you've already computed those values in the last `N` iterations. I think it should be stored in the layer in the transformed number space, not in the raw space and later to be multiplied / added.

Reply:
Added default values and error messaging if base/multiplier are not meet them in https://github.com/ros-planning/navigation2/pull/1882/files#diff-f67a16397bb5979b7cbf4729ea53a451R88